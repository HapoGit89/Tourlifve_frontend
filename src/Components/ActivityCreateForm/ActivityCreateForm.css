
.ActivityCreateForm{
    display: flex;
    flex-direction: row;
    justify-content: space-around;
    align-items: center;
    margin-top: 5vh;
    font-family: myFont;
}



.ActivityMap{
    margin: 2%
}

.results {
    margin-top: 5vh;
    margin-bottom: 5vh;
    display: flex;
    flex-direction: column;
    justify-content: space-between;
}


@font-face {
    font-family: myFont;
    src: url("../../fonts/Ronysiswadi9Light-DO5K9.ttf");
  }


  @media only screen and (max-width: 1000px){
    .ActivityCreateForm{
        display: flex;
        flex-direction: column;
        justify-content: space-around;
        align-items: center;
        margin-top: 13vh;
        margin-bottom: 5vh;
        font-family: myFont;
    }

    .results {
        display: flex;
        flex-direction: column;
        align-items: center;
        margin-bottom: 5vh;
    }
  }