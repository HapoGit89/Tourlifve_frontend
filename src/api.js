import axios from "axios";

const BASE_URL = process.env.REACT_APP_BASE_URL || "http://localhost:3001";

/** API Class.
 *
 * Static class providing API functionality.
 */

export class TourApi {
  // store user token for API access
  static token = localStorage.getItem("token") || ""


// universal request method
  static async request(endpoint, data = {}, method = "get") {
    console.debug("API Call:", endpoint, data, method);

    const url = `${BASE_URL}/${endpoint}`;
    const headers = { Authorization: `Bearer ${this.token}` ,  'Access-Control-Allow-Origin': '*',};
    const params = (method === "get")
        ? data
        : {};

    try {
      return (await axios({ url, method, data, params, headers })).data;
    } catch (err) {
      
      return err
      // let message = err.response.data.error.message;
      // throw Array.isArray(message) ? message : [message];
    }
  }

  // Individual API routes

  static async SignUp(data){
    try{
      let res = await this.request(`auth/register`, {...data}, "post")
      return res
      }
    catch(e){
      return (e)
    }
  }

  static async Login(data, callback){
    try{let res = await this.request(`auth/token`, {...data}, "post")
    return res}
    catch(e){
      return (e)
    }
}

static async getUser(username){
  try{
    let res = await this.request(`users/${username}`)
    return res
  }
  catch(e){
    return e
  }
}

static async deleteUser(username){
  try{
    let res = await this.request(`users/${username}`,{}, "delete")
    return res
  }
  catch(e){
    return e
  }
}

static async patchUser(username,data){
  try{
    let res = await this.request(`users/${username}`,{...data}, "patch")
    return res
  }
  catch(e){
    return e
  }
}

static async postTour(data){
  try{
    let res = await this.request(`tours/`,{...data},"post")
    return res
  }
  catch(e){
    return e
  }
}

static async getTourDetails(id){
  try{
    let res = await this.request(`tours/${id}`,{},"get")
    return res
  }
  catch(e){
    return e
  }
}

static async getTourstopDetails(id){
  try{
    let res = await this.request(`tourstops/${id}`,{},"get")
    return res
  }
  catch(e){
    return e
  }
}

static async patchTour(id, data){
  try{
    let res = await this.request(`tours/${id}`,{...data},"patch")
    return res
  }
  catch(e){
    return e
  }
}

}

// for now, put token ("testuser" / "password" on class)
// JoblyApi.token = "eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJ1c2VybmFtZ" +
//     "SI6InRlc3R1c2VyIiwiaXNBZG1pbiI6ZmFsc2UsImlhdCI6MTU5ODE1OTI1OX0." +
//     "FtrMwBQwe6Ue-glIFgz_Nf8XxRT2YecFCiSpYL0fCXc";