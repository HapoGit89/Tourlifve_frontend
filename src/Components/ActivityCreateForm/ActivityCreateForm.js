
import { useContext, useState, useEffect } from "react";
import { useParams, useNavigate } from "react-router-dom";
import ActivityMap from "../ActivityMap/ActivityMap";
import PoiResultCard from "../PoiResultCard/PoiResultCard";
import ActivitySearchForm from "../ActivitySearchForm/ActivitySearchForm";
import { fromPlaceId, setKey} from "react-geocode";
import { TourApi } from "../../api";
import userContext from "../../userContext";
import "./ActivityCreateForm.css"


function ActivityCreateForm(data) {
  // React controlled Form for User Login
  const user = useContext(userContext)
  const [tourstop, setTourstop] = useState()
  const [results, setResults] = useState({})
  const [selected, setSelected] = useState({})
  const {tourstop_id} = useParams()
  const navigate = useNavigate()
  setKey(process.env.REACT_APP_API_KEY)

   // get tourstop info from TourApi and store in state
  useEffect(()=>{
    const getTourstopData= async()=>{
        const res = await TourApi.getTourstopDetails(tourstop_id) 
        setTourstop(res.tourstop)
        }
        getTourstopData()
  },[user])


const handleSearch = async (data) => {
      const res = await TourApi.searchPlaces({lat: tourstop.lat, lng: tourstop.lng, origin_id: tourstop.googleplaces_id, mode: data.mode, duration: data.traveltime, query:data.keyword})
      const geoPromises = []
      for(let i = 0; i < res.destinations.length; i++){
        geoPromises.push(fromPlaceId(res.destinations[i].place_id))
      }
      Promise.all(geoPromises).then(results=>{
            for(let i=0; i <results.length; i++){
          res.destinations[i].position= results[i].results[0].geometry.location
            
            }
            if(res.destinations.length == 0){
              alert("Sorry couldnt find any places for your search!")
            }
            setResults(res.destinations)
      })
}



const handleMarkerClick = (id)=>{
  const selectedData = results.filter((el)=>el.place_id===id)
  setSelected(selectedData[0])
}



const createActivity = async () => {

          let poi = {}
        // post location or get existing location from db
        const res = await TourApi.postPoi({name: selected.name,
            address: selected.address, 
            googleplaces_id: selected.place_id, 
            googlemaps_link: selected.googlemaps_uri,
            category: selected.category || ""})


        if(res.poi){
        poi = res.poi
        }

        else if(res.response.data.error.message.slice(0,13)==="Duplicate poi"){
        const res2 = await TourApi.getAllPois()
        poi = res2.pois.filter((el)=>el.googleplaces_id==selected.place_id)[0]
        }

        else {
        alert ( 'ooops something went wrong, please try again or contact support')
        }

        const res3 = await TourApi.postActivity({tourstop_id: Number(tourstop_id), poi_id: poi.id, traveltime: selected.duration_value_secs, travelmode: selected.mode})

        if (res3.activity){
        alert(`Created Activity at ${selected.name} for tourstop ${tourstop_id}`)
        navigate("./../")
        }
        else if ( res3.response.data.error.message.slice(0,18)==="Duplicate activity"){
        alert("Sorry that activity already exists!")
        }
        else {
        alert("oops something went wrong")
        }
}



  if (user.token && tourstop) {
    return (<div className="ActivityCreateForm">

       

    <div className="ActivityMap">
            <ActivityMap handleMarkerClick={handleMarkerClick} key={results.length} results={results} location = {{location: {lat: tourstop.lat, lng: tourstop.lng}, name:tourstop.name}}></ActivityMap>
  
    </div>

   

      <div className="results">
      <PoiResultCard createActivity={createActivity}key={selected.place_id} activity={selected}></PoiResultCard>
      <ActivitySearchForm handleSearch={handleSearch} ></ActivitySearchForm>
       
        

      </div>

    </div> 





    )
  }

  else {
    return (
      <div>
        <h1>Sorry this only available for logged in users</h1>
      </div>
    )
  }
}


export default ActivityCreateForm