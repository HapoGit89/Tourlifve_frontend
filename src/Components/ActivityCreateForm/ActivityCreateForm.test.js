import { screen, render, waitFor, fireEvent } from '@testing-library/react';
import ActivityCreateForm from './ActivityCreateForm';
import userContext from '../../userContext';
import { BrowserRouter } from 'react-router-dom';
import '@testing-library/jest-dom'
import { act } from 'react-test-renderer';
import { TourApi } from '../../api';


const tourstop = {
  id: 10,
  name: "Alte Feuerwache Mannheim",
  country: "Germany",
  city: "Mannheim",
  street: "Brückenstraße",
  housenumber: "2",
  googleplaces_id: "ChIJ_9zn2CjMl0cR89fkHTMd_3c",
  tour_id: 7,
  location_id: 7,
  date: "2024-02-26T00:00:00.000Z",
  lat: 49.4957438,
  lng: 8.474236,
  activities: [
    ]
}


// Smoke Test

it('renders without crashing', () => {
    const user = {username: "hannes", token: "ajshajhjshajhjhs"}
    render( <userContext.Provider value={user}>
           <BrowserRouter>
          <ActivityCreateForm></ActivityCreateForm>
           </BrowserRouter>
           </userContext.Provider>
      )
});


// Snapshot Test

it('matches Snapshot', () => {
    const user = {username: "hannes", token: "ajshajhjshajhjhs"}
    const {asFragement} = render(<BrowserRouter>
            <userContext.Provider value={user}>
            <ActivityCreateForm></ActivityCreateForm>
            </userContext.Provider>
           </BrowserRouter>
      )
      expect(asFragement).toMatchSnapshot()
});


// specialized Tests

it('renders Search Form and Form', async() => {
  const ts = tourstop
    TourApi.getTourstopDetails = jest.fn()

    TourApi.getTourstopDetails.mockReturnValue({tourstop: ts})
  
  const user = {username: "Hannes", token: "askakjskjskjs"}
  const {getByText, getByPlaceholderText }=act(()=>{render(
          <userContext.Provider value={user}>
          <BrowserRouter>
          <ActivityCreateForm/>
          </BrowserRouter>
          </userContext.Provider>)})

          await waitFor(()=>expect(screen.getByText("Search For Interesting Places:")).toBeInTheDocument())
        
});

it('SearchForm Fires API Call', async() => {
  const ts = tourstop
    TourApi.getTourstopDetails = jest.fn()
    TourApi.searchPlaces = jest.fn()
    TourApi.searchPlaces.mockReturnValue({destinations: []})
    TourApi.getTourstopDetails.mockReturnValue({tourstop: ts})
  
  const user = {username: "Hannes", token: "askakjskjskjs"}
  const {getByText, getByPlaceholderText }=act(()=>{render(
          <userContext.Provider value={user}>
          <BrowserRouter>
          <ActivityCreateForm/>
          </BrowserRouter>
          </userContext.Provider>)})

          await waitFor(()=>expect(screen.getByText("Search For Interesting Places:")).toBeInTheDocument())
          const keyword = screen.getByPlaceholderText("Keyword...")
          const time = screen.getByLabelText("Traveltime")
          const mode= screen.getByLabelText("Mode of Travel:")
          const searchButton = screen.getByTestId("SearchButton")
          
          expect(keyword).toBeInTheDocument()
          expect(time).toBeInTheDocument()
          expect(mode).toBeInTheDocument()
          expect(searchButton).toBeInTheDocument()
          fireEvent.input(keyword, {target: {value: "Restaurant"}})
          fireEvent.change(mode, {target: {value: "walking"}})
          fireEvent.change(time, {target: {value: "5 min"}})
          expect(TourApi.getTourstopDetails).toHaveBeenCalled()
          expect(keyword).toHaveDisplayValue("Restaurant")
          expect(mode).toHaveDisplayValue("walking")
          expect(time).toHaveDisplayValue("5 min")
          fireEvent.click(searchButton)
          expect(TourApi.searchPlaces).toHaveBeenCalledTimes(1)

});


