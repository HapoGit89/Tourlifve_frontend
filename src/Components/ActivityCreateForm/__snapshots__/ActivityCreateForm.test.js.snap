// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[` 1`] = `undefined`;

exports[`Renders Searchform 1`] = `undefined`;

exports[`matches Snapshot 1`] = `undefined`;

exports[`matches Snapshot 2`] = `undefined`;
